javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/ignore/trie/TrieImpl.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.tooling.ignore.trie;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+final class TrieImpl<V> implements Trie<V> {
+
+  private final Node<V> root;
+
+  private TrieImpl(Node<V> root) {
+    this.root = root;
+  }
+
+  @Override
+  public V getOrNull(CharSequence str) {
+    Node<V> node = root;
+    V lastMatchedValue = null;
+
+    for (int i = 0; i < str.length(); ++i) {
+      char c = str.charAt(i);
+      Node<V> next = node.getNext(c);
+      if (next == null) {
+        return lastMatchedValue;
+      }
+      node = next;
+      // next node matched, use its value if it's defined
+      lastMatchedValue = next.value != null ? next.value : lastMatchedValue;
+    }
+
+    return lastMatchedValue;
+  }
+
+  static final class Node<V> {
+    final char[] chars;
+    final Node<V>[] children;
+    final V value;
+
+    Node(char[] chars, Node<V>[] children, V value) {
+      this.chars = chars;
+      this.children = children;
+      this.value = value;
+    }
+
+    @Nullable
+    Node<V> getNext(char c) {
+      int index = Arrays.binarySearch(chars, c);
```

Comment:
If you feel like comparing some optimizations, some ideas are

- Don't binary search, these arrays seem small enough that it would be slower than a normal loop
- A single array with every other element `Character` and `Node` - single array means better cache performance
- Copy in from Armeria (while simplifying it's funcionality for parameters) :P

https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/server/RoutingTrieBuilder.java

The main difference is the nodes are subpaths rather than characters - it's quite unfortunate a boring path like `org.springframework.blah` has a node per character